File: Sources/HedosKernel/Records/ModelRecord.swift
import CryptoKit
import Foundation

public struct ModelSource: Codable, Hashable, Sendable {
    public var kind: SourceKind
    public var path: String
    public var repo: String?
    public var ref: String?

    public init(kind: SourceKind, path: String, repo: String? = nil, ref: String? = nil) {
        self.kind = kind
        self.path = path
        self.repo = repo
        self.ref = ref
    }

    public var identity: String {
        "\(kind.rawValue)|\(path)|\(repo ?? "")"
    }
}

public struct RuntimeRef: Codable, Hashable, Sendable {
    public enum Resolution: String, Codable, Hashable, Sendable {
        case auto
        case user
        case unresolved
    }

    public var id: RuntimeID?
    public var resolved: Resolution
    public var tier: RunTier
    public var alternatives: [RuntimeID]
    public var confirmedAt: Date?

    public init(
        id: RuntimeID? = nil,
        resolved: Resolution = .unresolved,
        tier: RunTier = .recipeNeeded,
        alternatives: [RuntimeID] = [],
        confirmedAt: Date? = nil
    ) {
        self.id = id
        self.resolved = resolved
        self.tier = tier
        self.alternatives = alternatives
        self.confirmedAt = confirmedAt
    }

    public static let unresolved = RuntimeRef()
}

public struct ParamSpec: Codable, Hashable, Sendable {
    public enum ParamType: String, Codable, Hashable, Sendable {
        case int
        case float
        case bool
        case string
        case enumeration = "enum"
    }

    public var key: String
    public var type: ParamType
    public var defaultValue: JSONValue?
    public var range: [JSONValue]?
    public var values: [String]?

    public init(
        key: String,
        type: ParamType,
        defaultValue: JSONValue? = nil,
        range: [JSONValue]? = nil,
        values: [String]? = nil
    ) {
        self.key = key
        self.type = type
        self.defaultValue = defaultValue
        self.range = range
        self.values = values
    }

    enum CodingKeys: String, CodingKey {
        case key, type, range, values
        case defaultValue = "default"
    }
}

public struct ModelRecord: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var name: String
    public var modality: Modality
    public var capabilities: [Capability]
    public var source: ModelSource
    public var runtime: RuntimeRef
    public var params: [ParamSpec]
    public var paramValues: [String: JSONValue]
    public var systemPrompt: String?
    public var alias: String?
    public var execution: ExecutionMode
    public var footprintMB: Int?
    public var state: ModelState
    public var registeredAt: Date
    public var primaryWeightPath: String?
    public var contextLength: Int?
    public var hasChatTemplate: Bool?
    public var stopTokens: [String]?

    public init(
        name: String,
        modality: Modality,
        capabilities: [Capability],
        source: ModelSource,
        runtime: RuntimeRef = .unresolved,
        params: [ParamSpec] = [],
        paramValues: [String: JSONValue] = [:],
        systemPrompt: String? = nil,
        alias: String? = nil,
        execution: ExecutionMode = .sync,
        footprintMB: Int? = nil,
        state: ModelState = .unresolved,
        registeredAt: Date = Date(),
        contextLength: Int? = nil,
        hasChatTemplate: Bool? = nil,
        stopTokens: [String]? = nil
    ) {
        self.id = Self.stableID(for: source)
        self.name = name
        self.modality = modality
        self.capabilities = capabilities
        self.source = source
        self.runtime = runtime
        self.params = params
        self.paramValues = paramValues
        self.systemPrompt = systemPrompt
        self.alias = alias
        self.execution = execution
        self.footprintMB = footprintMB
        self.state = state
        self.registeredAt = registeredAt
        self.contextLength = contextLength
        self.hasChatTemplate = hasChatTemplate
        self.stopTokens = stopTokens
    }

    enum CodingKeys: String, CodingKey {
        case id, name, modality, capabilities, source, runtime, params, paramValues
        case systemPrompt, alias, execution, footprintMB, state, registeredAt
        case primaryWeightPath
        case contextLength, hasChatTemplate, stopTokens
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decode(String.self, forKey: .id)
        self.name = try container.decode(String.self, forKey: .name)
        self.modality = try container.decode(Modality.self, forKey: .modality)
        self.capabilities = try container.decode([Capability].self, forKey: .capabilities)
        self.source = try container.decode(ModelSource.self, forKey: .source)
        self.runtime = try container.decode(RuntimeRef.self, forKey: .runtime)
        self.params = try container.decode([ParamSpec].self, forKey: .params)
        self.paramValues =
            try container.decodeIfPresent([String: JSONValue].self, forKey: .paramValues) ?? [:]
        self.systemPrompt = try container.decodeIfPresent(String.self, forKey: .systemPrompt)
        self.alias = try container.decodeIfPresent(String.self, forKey: .alias)
        self.execution = try container.decode(ExecutionMode.self, forKey: .execution)
        self.footprintMB = try container.decodeIfPresent(Int.self, forKey: .footprintMB)
        self.state = try container.decode(ModelState.self, forKey: .state)
        self.registeredAt = try container.decode(Date.self, forKey: .registeredAt)
        self.primaryWeightPath = try container.decodeIfPresent(
            String.self, forKey: .primaryWeightPath)
        self.contextLength = try container.decodeIfPresent(Int.self, forKey: .contextLength)
        self.hasChatTemplate = try container.decodeIfPresent(Bool.self, forKey: .hasChatTemplate)
        self.stopTokens = try container.decodeIfPresent([String].self, forKey: .stopTokens)
    }

    public var displayName: String {
        guard let alias, !alias.isEmpty else { return name }
        return alias
    }

    public static func stableID(for source: ModelSource) -> String {
        let digest = SHA256.hash(data: Data(source.identity.utf8))
        return digest.prefix(8).map { String(format: "%02x", $0) }.joined()
    }
}

File: kernel/src/records/model_record.rs
//! The central data model: a `ModelRecord` and the value types it composes.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::records::identifiers::{
    Capability, ExecutionMode, Modality, ModelState, RunTier, RuntimeId, SourceKind,
};
use crate::records::json_value::JsonValue;
use crate::util::{hex_encode, now_millis};

/// Where a model's weights live and how it is identified.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct ModelSource {
    /// The kind of store the model came from.
    pub kind: SourceKind,
    /// The on-disk path (or endpoint identifier) of the model.
    pub path: String,
    /// The Hugging Face / Ollama repository, when applicable.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub repo: Option<String>,
    /// The revision/ref within the repository, when applicable.
    #[serde(rename = "ref", default, skip_serializing_if = "Option::is_none")]
    pub reference: Option<String>,
}

impl ModelSource {
    /// Create a source with just a kind and path.
    pub fn new(kind: SourceKind, path: &str) -> Self {
        Self {
            kind,
            path: path.to_owned(),
            repo: None,
            reference: None,
        }
    }

    /// The identity string a stable id is derived from: `kind|path|repo`.
    pub fn identity(&self) -> String {
        format!(
            "{}|{}|{}",
            self.kind.as_str(),
            self.path,
            self.repo.as_deref().unwrap_or("")
        )
    }
}

/// How a model was resolved to its runtime.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Resolution {
    /// Chosen automatically by the runtime auction.
    Auto,
    /// Pinned by the user.
    User,
    /// Not yet resolved.
    #[default]
    Unresolved,
}

/// The runtime a model resolves to, plus how that choice was made.
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct RuntimeRef {
    /// The winning runtime, if resolved.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<RuntimeId>,
    /// How the runtime was chosen.
    #[serde(default)]
    pub resolved: Resolution,
    /// How much support the runtime needs.
    #[serde(default)]
    pub tier: RunTier,
    /// Other runtimes that could also serve this model.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub alternatives: Vec<RuntimeId>,
    /// When the user confirmed the runtime, epoch millis.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub confirmed_at: Option<i64>,
}

/// The type of a tunable parameter.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ParamType {
    /// An integer.
    Int,
    /// A floating-point number.
    Float,
    /// A boolean.
    Bool,
    /// A free string.
    String,
    /// One of a fixed set of string values.
    Enum,
}

/// The schema of one tunable parameter a model exposes.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ParamSpec {
    /// The parameter's key, e.g. `temperature`.
    pub key: String,
    /// The parameter's type.
    #[serde(rename = "type")]
    pub param_type: ParamType,
    /// The default value, if any.
    #[serde(rename = "default", default, skip_serializing_if = "Option::is_none")]
    pub default_value: Option<JsonValue>,
    /// The `[min, max]` range for numeric parameters, if any.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub range: Option<Vec<JsonValue>>,
    /// The allowed values for an enum parameter, if any.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub values: Option<Vec<String>>,
}

/// A single model the kernel knows about: what it is, where it lives, how it runs,
/// and how the user has configured it.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ModelRecord {
    /// Stable identity derived from the source (see [`stable_id`]).
    pub id: String,
    /// The model's display name.
    pub name: String,
    /// Its primary modality.
    pub modality: Modality,
    /// Everything it can be asked to do.
    pub capabilities: Vec<Capability>,
    /// Where its weights live.
    pub source: ModelSource,
    /// The runtime it resolves to.
    #[serde(default)]
    pub runtime: RuntimeRef,
    /// The parameter schema it exposes.
    #[serde(default)]
    pub params: Vec<ParamSpec>,
    /// User-set parameter values.
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub param_values: BTreeMap<String, JsonValue>,
    /// A user-set system prompt override.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub system_prompt: Option<String>,
    /// A user-set display alias.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub alias: Option<String>,
    /// How the runtime delivers output.
    #[serde(default)]
    pub execution: ExecutionMode,
    /// Estimated memory footprint in megabytes.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub footprint_mb: Option<i64>,
    /// The record's lifecycle state.
    #[serde(default)]
    pub state: ModelState,
    /// When the record was first registered, epoch millis.
    pub registered_at: i64,
    /// The primary weight file, when one file dominates.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub primary_weight_path: Option<String>,
    /// The model's context window, when known.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context_length: Option<i64>,
    /// Whether the model ships a chat template.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub has_chat_template: Option<bool>,
    /// Stop tokens the model declares.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub stop_tokens: Option<Vec<String>>,
    /// Whether the model is still downloading.
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub downloading: bool,
    /// A content fingerprint used to follow a model across moves.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub content_fingerprint: Option<String>,
}

impl ModelRecord {
    /// Create a record for a model, deriving its stable id from `source` and
    /// stamping `registered_at` with the current time.
    pub fn new(
        name: &str,
        modality: Modality,
        capabilities: Vec<Capability>,
        source: ModelSource,
    ) -> Self {
        Self {
            id: stable_id(&source),
            name: name.to_owned(),
            modality,
            capabilities,
            source,
            runtime: RuntimeRef::default(),
            params: Vec::new(),
            param_values: BTreeMap::new(),
            system_prompt: None,
            alias: None,
            execution: ExecutionMode::Sync,
            footprint_mb: None,
            state: ModelState::Unresolved,
            registered_at: now_millis(),
            primary_weight_path: None,
            context_length: None,
            has_chat_template: None,
            stop_tokens: None,
            downloading: false,
            content_fingerprint: None,
        }
    }

    /// The name to show the user: the alias if set, otherwise the name.
    pub fn display_name(&self) -> &str {
        match &self.alias {
            Some(alias) if !alias.is_empty() => alias,
            _ => &self.name,
        }
    }

    /// Whether the model can perform `capability`.
    pub fn can(&self, capability: &Capability) -> bool {
        self.capabilities.contains(capability)
    }
}

/// The stable identity of a model: the first eight bytes of the SHA-256 of the
/// source's `(kind, path, repo)` fields, hex-encoded. The fields are hashed with
/// a length prefix so the mapping is injective — a `|` (or any byte) inside a
/// path or repo cannot make two distinct sources collide. Two sources with the
/// same kind, path, and repo map to the same id.
pub fn stable_id(source: &ModelSource) -> String {
    let mut hasher = Sha256::new();
    for field in [
        source.kind.as_str(),
        source.path.as_str(),
        source.repo.as_deref().unwrap_or(""),
    ] {
        hasher.update((field.len() as u64).to_le_bytes());
        hasher.update(field.as_bytes());
    }
    let digest = hasher.finalize();
    hex_encode(&digest[..8])
}
